### DISCUSSION TITLE
ID Documents Detection and Classification with YOLOv8 and Orientation Correction

### DISCUSSION THREAD

**Discussion by MayThazinPhyo** [ultralytics/ultralytics#13469] [Q&A]
Hi everyone,
I want to detect the different id cards and classify the classes like passport or other types of ID card. I trained yolov8 models on custom dataset and it works well on correct orientated images but not in images which are oriented vertically. Should I trained vertically oriented images as well to get detected or any other suggestions?

**Comment by glenn-jocher** [comment]
Hi there!

It's great to hear that you've successfully trained a YOLOv8 model for detecting and classifying ID cards! For handling vertically oriented images, you have a couple of options:

1. **Augment Your Training Data**: Including vertically oriented images in your training dataset can significantly improve your model's ability to detect and classify ID cards in various orientations. You can use data augmentation techniques to rotate your images during training. Here's an example of how you can do this using the `albumentations` library:

    ```python
    import albumentations as A
    from albumentations.pytorch import ToTensorV2

    transform = A.Compose([
        A.Rotate(limit=90, p=0.5),  # Randomly rotate images by up to 90 degrees
        A.HorizontalFlip(p=0.5),    # Randomly flip images horizontally
        A.VerticalFlip(p=0.5),      # Randomly flip images vertically
        A.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
        ToTensorV2(),
    ])
    ```

    This will help your model generalize better to different orientations.

2. **Post-Processing for Orientation Correction**: Another approach is to handle orientation correction as a post-processing step. You can use an additional model or heuristic to detect the orientation of the ID card and then rotate it to the correct orientation before passing it to your YOLOv8 model for detection and classification.

    For example, you could use OpenCV to detect the orientation and correct it:

    ```python
    import cv2

    def correct_orientation(image):
        # Detect edges
        edges = cv2.Canny(image, 50, 150)
        # Find contours
        contours, _ = cv2.findContours(edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        # Assume the largest contour is the ID card
        contour = max(contours, key=cv2.contourArea)
        # Get the rotated bounding box
        rect = cv2.minAreaRect(contour)
        angle = rect[-1]
        # Correct the angle
        if angle < -45:
            angle = -(90 + angle)
        else:
            angle = -angle
        # Rotate the image
        (h, w) = image.shape[:2]
        center = (w // 2, h // 2)
        M = cv2.getRotationMatrix2D(center, angle, 1.0)
        rotated = cv2.warpAffine(image, M, (w, h))
        return rotated
    ```

    This function will rotate the image to the correct orientation before detection.

3. **Ensure Latest Versions**: Make sure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:

    ```bash
    pip install --upgrade torch ultralytics
    ```

    This ensures you benefit from the latest features and bug fixes.

If you need further assistance, please provide a minimum reproducible code example so we can better understand your setup and help you more effectively. You can find more details on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Hope this helps! Feel free to reach out if you have any more questions 😊
